### ISSUE TITLE
format constraint in input_descriptor

### ISSUE THREAD

**Issue by tlodderstedt** [decentralized-identity/presentation-exchange#242]
I think there is a need for formats at two different levels:

capabilities: what format are verifiers/holders able to process
credential constraints: what format expects a verifier a certain credential to be represented in.
(1) can be represented in OIDC using client/OP metadata or the way PE defines it
(2) is missing in PE - it is however very important since it influences the way paths are defined, e.g. $.credentialSubject.account[].id seems to be a path within a JSON-LD-based credential whereas $.vc.credentialSubject.account[].id seems to be referring to an element within a JWT-based credential. Using them in parallel in constraints (see https://identity.foundation/presentation-exchange/#input-descriptor) to me seems like a very error prone reference mechanism.

I think adding format on the level of the input_descriptor would improve PE's robustness significantly.

**Comment by tlodderstedt** [comment]
@David-Chadwick I suggested to add a constraint to the input descriptor to determine the credential format the verifier wants to receive. If there is a credential in the user's wallet with the correct schema/type but a different format, that credential simply does not fit the input descriptor so the wallet can reject the request or suggest the end-user to obtain a suitable credential. 
The format element in the presentation definition might be sufficient in some cases, however the ability to process a certain format does not mean the intend and will to process any requested credential in any format. Even more important, I would like to see a way to specify constraints over credentials in a robust way. path constraints, esp., are dependent on the credential formats. The fuzzy matching logic implied (or shall I say forced) by the current spec is not robust.

**Comment by David-Chadwick** [comment]
> @David-Chadwick I suggested to add a constraint to the input descriptor to determine the credential format the verifier wants to receive.

My understanding is that this is precisely the purpose of presentationDefinition.format. So there is no need for it in inputDescriptor

> If there is a credential in the user's wallet with the correct schema/type but a different format, that credential simply does not fit the input descriptor so the wallet can reject the request or suggest the end-user to obtain a suitable credential.

The credential would not fit the presentationDefinition.format and therefore would not be selected.

> The format element in the presentation definition might be sufficient in some cases, however the ability to process a certain format does not mean the intend and will to process any requested credential in any format.

I think we need the specifiers of presentationDefinition.format to say precisely what the intent of the presentationDefinition.format definition is. From my perspective as an implementor, if my Verifier can accept jwt formatted VCs, then it can process all types/schemas that it supports in jwt format. It does not make sense (to me) to accept VCs of one type/schema with JSON-LD proofs and of another type/schema with jwt signatures.



> Even more important, I would like to see a way to specify constraints over credentials in a robust way. path constraints, esp., are dependent on the credential formats.

That is certainly true for VCs and OIDC claims. But it is not true for VCs of any proof format as stated in my original comment. This is because the credentials of all VCs should be in the identical format that is specified in the W3C Data Model. There are not different credential formats in the W3C recommendation. There is only one.

> The fuzzy matching logic implied (or shall I say forced) by the current spec is not robust.

This is in part due to the alternative ways that certain fields can be specified in the W3C VC recommendation. The issuer field can be specified as follows "The value of the issuer property MUST be either a URI or an object containing an id property." Thus matching the issuer id using JSON schemas requires two alternative paths, which is a pain!

**Comment by tlodderstedt** [comment]
> > @David-Chadwick I suggested to add a constraint to the input descriptor to determine the credential format the verifier wants to receive.
> 
> My understanding is that this is precisely the purpose of presentationDefinition.format. So there is no need for it in inputDescriptor

A single presentation definition may contain multiple input descriptors. I basically don't buy into the assumption a verifier will be able to process every presentations/credentials in any possible format since credentials in different formats require different processing logic. Some need unmarshalling, others scheme/namespace/context handling. Signature validation differs, paths to end-user claims differ.

The current syntax forces implementation not willed to support the cartesian product from formats and schema defined in a presentation definition to use separate presentation definitions, each for the format specific input descriptors. This also means, a verifier cannot define submission requirements for those different descriptors.

For me, the most natural (and most flexible) way is to specify the format as constraint on the input descriptor level. This also allows to use the exact format specific path for field constraints. 

This is a much more reliable way to define constraints than listing several possible paths to a certain kind of claim. Just take a look at the example at https://identity.foundation/presentation-exchange/spec/v1.0.0/#input-descriptor, esp. 

```
          "path": [
                "$.credentialSubject.account[*].id",
                "$.vc.credentialSubject.account[*].id",
                "$.account[*].id"
              ],
```

I'm concerned what happens if the wallet contains a credential with the correct schema but a format the verifier did not expect for that kind of credential and the path constraint coincidentally matches an id claim with a completely different semantics. In worst case, this might result in a security hole. 

> 
> > If there is a credential in the user's wallet with the correct schema/type but a different format, that credential simply does not fit the input descriptor so the wallet can reject the request or suggest the end-user to obtain a suitable credential.
> 
> The credential would not fit the presentationDefinition.format and therefore would not be selected.
> 
> > The format element in the presentation definition might be sufficient in some cases, however the ability to process a certain format does not mean the intend and will to process any requested credential in any format.
> 
> I think we need the specifiers of presentationDefinition.format to say precisely what the intent of the presentationDefinition.format definition is. From my perspective as an implementor, if my Verifier can accept jwt formatted VCs, then it can process all types/schemas that it supports in jwt format. It does not make sense (to me) to accept VCs of one type/schema with JSON-LD proofs and of another type/schema with jwt signatures.

Have you seen this in a productive application? As stated above, I don’t believe a verifier will support the cartesian product of all types and formats for all credentials. 

> 
> > Even more important, I would like to see a way to specify constraints over credentials in a robust way. path constraints, esp., are dependent on the credential formats.
> 
> That is certainly true for VCs and OIDC claims. But it is not true for VCs of any proof format as stated in my original comment. This is because the credentials of all VCs should be in the identical format that is specified in the W3C Data Model. There are not different credential formats in the W3C recommendation. There is only one.

really? The claims in JSON-LD credentials live underneath "credentialSubject", whereas JWT-based credentials use the path "vc.credentialSubject". The issuer is represented in the claims "issuer" and "iss", respectively. How is that identical?

> 
> > The fuzzy matching logic implied (or shall I say forced) by the current spec is not robust.
> 
> This is in part due to the alternative ways that certain fields can be specified in the W3C VC recommendation. The issuer field can be specified as follows "The value of the issuer property MUST be either a URI or an object containing an id property." Thus matching the issuer id using JSON schemas requires two alternative paths, which is a pain!

yep. And I want to see a robust way to cope with it instead of a cool and fuzzy mechanism opening up stability and security holes.

**Comment by csuwildcat** [comment]
PROPOSAL: add a `format` property inside Input Descriptors, the value of which must be the same object/syntax as defined in https://identity.foundation/presentation-exchange/spec/v1.0.0/#presentation-definition under the `format` property within it.

**Comment by csuwildcat** [comment]
This was addressed in the latest 2.0 PR #255
